tests/datasets/storages/processors/test_replaced_groups.py

```diff
@@ -127,3 +173,53 @@ def test_too_many_groups_to_exclude(query: ClickhouseQuery) -> None:
 
     assert query.get_condition() == build_in("project_id", [2])
     assert query.get_from_clause().final
+
+
+def test_query_overlaps_replacements_method(
+    query: ClickhouseQuery, query_with_timestamp: ClickhouseQuery,
+) -> None:
+    enforcer = PostReplacementConsistencyEnforcer("project_id", None)
+    # query from < latest replacement time < query to
+    assert enforcer._query_overlaps_replacements(
+        query_with_timestamp, datetime(2021, 1, 1, 10)
+    )
+    # query to < latest replacement time
+    assert enforcer._query_overlaps_replacements(
+        query_with_timestamp, datetime(2021, 2, 1)
+    )
+    # latest replacement time < query from
+    assert not enforcer._query_overlaps_replacements(
+        query_with_timestamp, datetime(2020, 1, 1)
+    )
+    # replacement time unknown, default to overlaps
+    assert enforcer._query_overlaps_replacements(query_with_timestamp, None)
+    # query time range unknown, default to overlaps
+    assert enforcer._query_overlaps_replacements(query, datetime(2021, 2, 1))
+
+
+def test_query_overlaps_replacements_processor(
+    query_with_timestamp: ClickhouseQuery, query_with_future_timestamp: ClickhouseQuery
+) -> None:
+    enforcer = PostReplacementConsistencyEnforcer("project_id", ReplacerState.EVENTS)
+
+    # replacement time unknown, default to "overlaps" but no groups to exclude so shouldn't be final
+    enforcer._set_query_final(query_with_timestamp, True)
+    enforcer.process_query(query_with_timestamp, HTTPRequestSettings())
+    assert not query_with_timestamp.get_from_clause().final
+
+    # overlaps replacement and should be final due to too many groups to exclude
+    state.set_config("max_group_ids_exclude", 2)
+    set_project_exclude_groups(
+        2,
+        [100, 101, 102],
+        ReplacerState.EVENTS,
+        ReplacementType.EXCLUDE_GROUPS,  # Arbitrary replacement type, no impact on tests
+    )
+    enforcer._set_query_final(query_with_timestamp, True)
```

Comment:
Please do not set the flag to True before the test. This way you are not testing the behavior of the `process_query` method anymore as you don't know if the final flag is on because it was set by the process_query method or by yourself on this line.